projects/main_ping_delays_pingsender.kp/knowledge.md

```diff
@@ -0,0 +1,387 @@
+
+
+```python
+pre_pings = Dataset.from_source("telemetry") \
+    .where(docType="main") \
+    .where(appUpdateChannel="nightly") \
+    .where(submissionDate=lambda x: "20170402" <= x < "20170408") \
+    .where(appBuildId=lambda x: "20170402" <= x < "20170408") \
+    .records(sc, sample=1)
+
+post_pings = Dataset.from_source("telemetry") \
+    .where(docType="main") \
+    .where(appUpdateChannel="nightly") \
+    .where(submissionDate=lambda x: "20170430" <= x < "20170506") \
+    .where(appBuildId=lambda x: "20170430" <= x < "20170506") \
+    .records(sc, sample=1)
+```
+To look at delays, we need to look at times. There are a lot of times, and they are recorded relative to different clocks.
+
+**`creationDate`** - The time the Telemetry code in Firefox created the ping, according to the client's clock, expressed as an ISO string. **`meta/creationTimestamp`** is the same time, but expressed in nanoseconds since the epoch.
+
+**`meta/Date`** - The time the Telemetry code in Firefox sent the ping to the server, according to the client's clock, expressed as a Date string conforming to [RFC 7231](https://tools.ietf.org/html/rfc7231#section-7.1.1.1).
+
+**`meta/Timestamp`** - The time the ping was received by the server, according to the server's
+clock, expressed in nanoseconds since the epoch.
+
+
+```python
+pre_subset = get_pings_properties(pre_pings, ["application/channel",
+                                              "id",
+                                              "creationDate",
+                                              "meta/creationTimestamp",
+                                              "meta/Date",
+                                              "meta/Timestamp",
+                                              "meta/X-PingSender-Version",
+                                              "payload/info/reason",
+                                              "payload/simpleMeasurements/shutdownDuration"])
+
+post_subset = get_pings_properties(post_pings, ["application/channel",
+                                                "id",
+                                                "creationDate",
+                                                "meta/creationTimestamp",
+                                                "meta/Date",
+                                                "meta/Timestamp",
+                                                "meta/X-PingSender-Version",
+                                                "payload/info/reason",
+                                                "payload/simpleMeasurements/shutdownDuration"])
+```
+The `shutdown` ping is a particular kind of `main` ping with the `reason` field set to `shutdown`, as it's saved during shutdown. Since there are other types of `main` pings, we need to filter undesired pings out. Moreover, we also create a filtered set of pings that are only sent using the `pingsender`.
```

Comment:
The filter for pingsender-only is way down the page. Probably shouldn't mention it this high. You can probably just limit this to the first sentence and be done with it.